torchx/schedulers/utils/ray_driver.py

```diff
@@ -0,0 +1,80 @@
+import os
+import json
+import dataclasses
+import ray
+from typing import Any, Dict, Iterable, List, Optional, Set, Type
+import logging
+import subprocess
+from ray_common import RayActor
+
+@ray.remote
+class CommandActor:
+    def __init__(self, command, env):
+        self.command = command
+        self.env = env
+
+    def run_command(self):
+        return subprocess.run(self.command, env = self.env)
+
+def load_actor_json(filename : str) -> List[Dict]:
+    with open(filename) as f:
+        actor = json.load(f)
+        actor = json.loads(actor)
+        return actor
+
+if __name__ == "__main__":
+
+    # Move to scheduler
+    actor1 = RayActor("resnet", ["echo", "hello resnet"])
+    actor2 = RayActor("bert", ["echo", "hello bert"])
+    serialize([actor1, actor2])
+
+    logging.basicConfig(filename='driver.log', encoding='utf-8', level=logging.DEBUG)
+
+
+    # On driver.py
+    logging.debug("Reading actor.json")
+    actors_dict = load_actor_json('actor.json')
+    pgs = []
+    ray.init(address="auto", namespace="torchx-ray")
+
+    for actor_dict in actors_dict:
+
+        bundle = {"CPU": actor_dict["num_cpus"], "GPU": actor_dict["num_gpus"]}
+        bundles = [bundle] * actor_dict["num_replicas"]
+        pg = ray.util.placement_group(bundles, strategy="SPREAD")
+        pgs.append(pg)
+
+        logging.debug("Waiting for placement group to start.")
+        ready = pg.wait(timeout_seconds=100)
+
+
+        if ready:
+            logging.debug("Placement group has started.")
+            logging.debug("Starting remote function")
+        
+        else:
+            raise TimeoutError(
+                "Placement group creation timed out. Make sure "
+                "your cluster either has enough resources or use "
+                "an autoscaling cluster. Current resources "
+                "available: {}, resources requested by the "
+                "placement group: {}".format(ray.available_resources(),
+                                            pg.bundle_specs))
+
+    actors = [CommandActor.options(placement_group=pgs[i], num_cpus=actor_dict[i]["num_cpus"], num_gpus=actor_dict[i]["num_gpus"]).remote(actors_dict[i]["command"], actor_dict["env"]) for i in range(len(actors_dict) * actor_dict["num_replicas"])]
+    ray.get([a.run_command.remote() for a in actors])
+    unfinished = [a.run_command.remote() for a in actors]
+
+    while len(unfinished) > 0:
+        finished, unfinished = ray.wait(unfinished)
+        # If a failure occurs the ObjectRef will be marked as finished.
+        # Calling ray.get will expose the failure as a RayActorError.
+        for object_ref in finished:
+            try:
+                status_code = ray.get(object_ref)
+            except ray.RayActorError as exc:
+                status_code = 1
+
+            if status_code != 0:
+                raise RuntimeError("Job failed")
```

Comment:
Is this the way we should report a job failure to Ray? Or is there a better approach? Can the user see `stdout` and `stderr` of the subprocesses running inside the remote actors? (Similar to how you see outputs of each process in a Slurm job)

Reply:
There's a couple points here:

1. For stdout and stderr, if we run the command in the main process instead of the subprocess (see the comment above about using `exec`), then the stdout and stderr from each actor will be shown on the driver side. Then, Ray Jobs will take the stdout and stderr from the driver (which contains the stdout and stderr from each actor) and persists it on the head node. Ray jobs exposes a `/logs` endpoint that torchX (or the user) can use to retrieve these logs. Does that work for you guys?
2. For exceptions, if an actor raises an error during the execution of the command, then this **same exception** will be raised when you do `ray.get(object_ref)`. This will then get propagated up to the driver, and will terminate the job and the exception message will be piped to stderr. Just as above, stderr can be accessed via the `/logs` endpoint that Ray Jobs exposes. 

Using `exec(...)` instead of `subprocess` should make both of the above much simpler.

Also, my fault for not clarifying this before, but a `RayActorError` will only be raised if the the object reference cannot be resolved (for example if the actor is pre-empted or a node goes down). In that case, we should catch the `RayActorError`, and either have some built-in retry logic or raise a more meaningful error